common/aws/ddb/state.go

```diff
@@ -238,149 +218,65 @@ func (sdb *StateDatabase) StoreCid(streamCid *models.StreamCid) (bool, error) {
 			},
-			ScanIndexForward: aws.Bool(false), // descending
-		},
-		func(streamCid *models.StreamCid) bool {
-			latest = streamCid
-			return false // stop iteration after the first entry
-		},
-	); err != nil {
-		return nil, err
-	}
-	return latest, nil
-}
-
-func (sdb *StateDatabase) UpdateAnchorTs(streamId, cid string, anchorTs time.Time) (bool, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), models.DefaultHttpWaitTime)
-	defer cancel()
-
-	updateItemIn := dynamodb.UpdateItemInput{
-		Key: map[string]types.AttributeValue{
-			"id":  &types.AttributeValueMemberS{Value: streamId},
-			"cid": &types.AttributeValueMemberS{Value: cid},
-		},
-		TableName:           aws.String(sdb.streamTable),
-		ConditionExpression: aws.String("attribute_not_exists(#ancTs)"),
-		ExpressionAttributeNames: map[string]string{
-			"#ancTs": "ats",
-		},
-		ExpressionAttributeValues: map[string]types.AttributeValue{
-			":ancTs": &types.AttributeValueMemberN{Value: strconv.FormatInt(anchorTs.Unix(), 10)},
-		},
-		UpdateExpression: aws.String("set #ancTs = :ancTs"),
-	}
-	if _, err := sdb.client.UpdateItem(ctx, &updateItemIn); err != nil {
-		// To get a specific API error
-		var condUpdErr *types.ConditionalCheckFailedException
-		if errors.As(err, &condUpdErr) {
-			// Not an error, just indicate that we couldn't update the entry
-			log.Printf("updateAnchorTs: anchor timestamp already set: %s, %s, %s, %v", streamId, cid, anchorTs, err)
-			return false, nil
+			Item:         attributeValues,
+			ReturnValues: types.ReturnValueAllOld,
 		}
-		log.Printf("updateAnchorTs: error writing to db: %v", err)
-		return false, err
-	}
-	return true, nil
-}
-
-func (sdb *StateDatabase) GetAnchoredCid(streamId, cid string) (*models.StreamCid, error) {
-	if streamCid, err := sdb.GetCid(streamId, cid); err != nil {
-		return nil, err
-	} else if streamCid.AnchorTs != nil {
-		return streamCid, nil
-	} else {
-		var anchoredCid *models.StreamCid = nil
-		if err = sdb.iterateCids(
-			&dynamodb.QueryInput{
-				TableName:              aws.String(sdb.streamTable),
-				IndexName:              aws.String(stateIdAnchorTsIndex),
-				KeyConditionExpression: aws.String("#id = :id and #ancTs >= :ts"),
-				ExpressionAttributeValues: map[string]types.AttributeValue{
-					":id": &types.AttributeValueMemberS{Value: streamId},
-					":ts": &types.AttributeValueMemberN{Value: strconv.FormatInt(streamCid.Timestamp.Unix(), 10)},
-				},
-				ExpressionAttributeNames: map[string]string{
-					"#id":    "id",
-					"#ancTs": "ats",
-				},
-				ScanIndexForward: aws.Bool(true), // ascending
-			},
-			func(streamCid *models.StreamCid) bool {
-				anchoredCid = streamCid
-				return false // stop iteration after the first entry
-			},
-		); err != nil {
-			return nil, err
+		if putItemOut, err := sdb.client.PutItem(ctx, &putItemIn); err != nil {
+			// To get a specific API error
+			var condUpdErr *types.ConditionalCheckFailedException
```

Comment:
This will happen if the condition specified in the `ConditionExpression` fails when attempting a write. For example, when updating a tip, the write can fail if the tip being written is older than the tip already in the DB.

Catching this error allows us to mark older tips REPLACED.